homework3/set_project_dir/set.h

```diff
@@ -0,0 +1,653 @@
+
+            _fix_node_height(v);
+
+            Node *cur_prev = cur->prev;
+            Node *cur_next = cur->next;
+
+            cur_prev->next = cur_next;
+            cur_next->prev = cur_prev;
+
+            _size--;
+
+            delete cur;
+            return;
+        }
+        if (cur->right == nullptr) {
+            v = cur->left;
+
+            _fix_node_height(v);
+
+            Node *cur_prev = cur->prev;
+            Node *cur_next = cur->next;
+
+            cur_prev->next = cur_next;
+            cur_next->prev = cur_prev;
+
+            _size--;
+
+            delete cur;
+            return;
+        }
+        if (_get_node_balance(cur) > 0) {
+            Node *to_swap = cur->right;
+            Node *to_swap_parent = cur;
+
+            while (to_swap->left != nullptr) {
+                to_swap_parent = to_swap;
+                to_swap = to_swap->left;
+            }
+
+            _node_swap(cur, p, to_swap, to_swap_parent);
+            _erase(v->right, key, v);
+            _fix_node_height(v);
+            return;
+        } else {
+            Node *to_swap = cur->left;
+            Node *to_swap_parent = cur;
+
+            while (to_swap->right != nullptr) {
+                to_swap_parent = to_swap;
+                to_swap = to_swap->right;
+            }
+
+            _node_swap(cur, p, to_swap, to_swap_parent);
+            _erase(v->left, key, v);
+            _fix_node_height(v);
+            return;
+        }
+    } else if (key < current_key) {
+        _erase(cur->left, key, cur);
+        _fix_node_height(cur);
+    } else {
+        _erase(cur->right, key, cur);
+        _fix_node_height(cur);
+    }
+    _balance_node(v);
+}
+
+template <class T>
+void Set<T>::_fix_node_height(Node *v) {
+    if (v == nullptr)
+        return;
+
+    if (v->right == nullptr && v->left == nullptr) {
+        v->height = 1;
+    } else if (v->left != nullptr) {
+        v->height = v->left->height + 1;
+    } else if (v->right != nullptr) {
+        v->height = v->right->height + 1;
+    } else {
+        v->height = std::max(v->left->height, v->right->height) + 1;
+    }
+}
+
+template <class T>
+int Set<T>::_get_node_balance(Node *v) {
+    if (v == nullptr)
+        return 0;
+
+    if (v->right == nullptr && v->left == nullptr) {
+        return 0;
+    } else if (v->right != nullptr) {
+        return v->right->height;
+    } else if (v->left != nullptr) {
+        return -(v->left->height);
+    } else {
+        return v->right->height - v->left->height;
+    }
+}
+
+template <class T>
+void Set<T>::_balance_node(Node *&v) {
+    if (v == nullptr)
+        return;
+
+    Node *cur = v;
+
+    if (_get_node_balance(cur) == -2) {
+        Node *cur_left = cur->left;
+        if (_get_node_balance(cur_left) > 0) {
+            Node *cur_left_right = cur_left->right;
+
+            v = cur_left_right;
+
+            cur->left = cur_left_right->right;
+            cur_left->right = cur_left_right->left;
```

Comment:
divide logic to sub private methods to simplify code

for instance: rotate_left , rotate_right and so on